### PR TITLE
Base version

### PR DESCRIPTION
**Pull request by RuslanProgrammer**

### CODE REVIEW

File: contracts/L1/CorePropertiesL1.sol [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    function setFee(address token_, uint256 fee_) external onlyOwner {
+        _fees[token_] = fee_;
+    }
+
+    function setTreasuryAddress(address treasuryAddress_) external onlyOwner {
+        treasuryAddress = treasuryAddress_;
+    }
+
+    function getFeeAndTreadury(address distributionAddress_) external view returns (uint256, address) {
```
Comment (tster) [inline contracts/L1/CorePropertiesL1.sol]: small typo here in function name, should be `getFeeAndTreasury`

**Comment by tster** [comment]
Noticed a typo throughout the codebase - `lzEnpointAddress` instead of `lzEndpointAddress`

Occurs in at least 8 places so won't list them all, Ctrl + F should do it!

File: contracts/L2/Mor20FactoryL2.sol [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import {RewardTokenDeployer} from "../libs/RewardTokenDeployer.sol";
+
+contract Mor20FactoryL2 is BaseFactoryL2, Ownable {
+    CorePropertiesL2 public coreProperties;
+
+    constructor(address coreProperties_) {
+        coreProperties = CorePropertiesL2(coreProperties_);
+    }
+
+    function deployMor20OnL1(Mor20DeployParams calldata parameters_) external onlyOwner {
```
Comment (tster) [inline contracts/L2/Mor20FactoryL2.sol:21]: shouldn't this method be called deployMor20On**L2**?

File: contracts/Factory.sol [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        _unpause();
+    }
+
+    /**
+     * @notice The function to set implementation for the specific pool.
+     *
+     * @param poolType_ the type of the pool.
+     * @param implementation_ the implementation the pool will point to.
+     */
+    function setImplementation(uint8 poolType_, address implementation_) public onlyOwner {
```
Comment (dovgopoly) [inline contracts/Factory.sol:44]: Should `implementation_` be beacon?

File: contracts/L2/L2Factory.sol [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            l2Params_.mor20Symbol,
+            lzExternalDeps.oftEndpoint,
+            _msgSender(),
+            l2MessageReceiver
+        );
+        deployedProxies[_msgSender()][l2Params_.protocolName][uint8(PoolType.MOR20)] = mor20;
+
+        IL2MessageReceiver(l2MessageReceiver).L2MessageReceiver__init(
+            mor20,
+            IL2MessageReceiver.Config(lzExternalDeps.endpoint, address(0), lzExternalDeps.senderChainId)
```
Comment (dovgopoly) [inline contracts/L2/L2Factory.sol:58]: Can we calculate L1Sender in advance?

File: contracts/L1/L1Factory.sol [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        IL1Sender.RewardTokenConfig memory lzConfig_ = IL1Sender.RewardTokenConfig(
+            lzExternalDeps.endpoint,
+            l1Params_.l2MessageReceiver,
+            lzExternalDeps.destinationChainId,
+            lzExternalDeps.zroPaymentAddress,
+            lzExternalDeps.adapterParams
+        );
+
+        IL1Sender.DepositTokenConfig memory arbConfig = IL1Sender.DepositTokenConfig(
```
Comment (dovgopoly) [inline contracts/L1/L1Factory.sol:70]:
```suggestion
        IL1Sender.DepositTokenConfig memory arbConfig_ = IL1Sender.DepositTokenConfig(
```

File: contracts/Factory.sol [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+pragma solidity ^0.8.20;
+
+import {ERC1967Proxy} from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {PausableUpgradeable} from "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+
+import {IFactory} from "./interfaces/IFactory.sol";
+
+abstract contract Factory is IFactory, OwnableUpgradeable, PausableUpgradeable, UUPSUpgradeable {
```
Comment (dovgopoly) [inline contracts/Factory.sol:11]: Let's add method to precalculate address. It should take deployer as a parameter.

**Review by dovgopoly — APPROVED** [review]
Sweet!

**Review by dovgopoly — APPROVED** [review]

**Pull request merged**